vmware-event-router/hack/run_integration_tests.sh

```diff
@@ -0,0 +1,130 @@
+
+cecho() {
+    local default_msg="No message passed."
+    message=${1:-$default_msg}
+    color=${2:-$RESET}
+
+    counter=$((${counter:-0} + 1))
+    echo -e ${color}"\n"\[ ${counter} \] ${message} ${RESET}
+    return
+}
+
+function cleanup() {
+    # capture previous return val, e.g. error
+    rv=$?
+    # don't run cleanup in CI env, e.g. Github Action sets CI=true
+    if [ "${CI:-no}" != true ]; then
+        cecho "Running cleanup and then exiting with code ${rv}" ${GREEN}
+        rm -rf ${tmp_dir} && rm -rf faas-cli-* && rm -rf template && ${KINDBIN} delete cluster --name ${KIND_CLUSTER}
+    fi
+    exit $rv
+}
+trap cleanup INT TERM EXIT
+
+# create kind cluster
+cecho "---> Creating Kubernetes cluster (${K8S_VERSION}) with max wait time: ${KIND_TIMEOUT}" ${GREEN}
+${TIMEOUT_CMD} ${KIND_TIMEOUT} ${KINDBIN} create cluster --name ${KIND_CLUSTER} --wait ${KIND_WAIT} --image ${KIND_IMAGE}
+
+# generate password used by OpenFaaS for basic_auth
+export OF_PASSWORD=$(head -c 12 /dev/urandom | shasum | cut -d' ' -f1)
+
+# if running as Github Actions make it available to other steps
+if [ "${CI:-no}" = true ]; then
+    echo "::set-env name=OF_PASSWORD::$OF_PASSWORD"
+fi
+
+# deploy faas-netes
+cecho "---> Deploying OpenFaaS (${OF_VERSION})" ${GREEN}
+git clone https://github.com/openfaas/faas-netes ${tmp_dir} && git -C ${tmp_dir} checkout ${OF_VERSION}
+kubectl create -f ${tmp_dir}/namespaces.yml
+kubectl -n openfaas create secret generic basic-auth --from-literal=basic-auth-user=admin --from-literal=basic-auth-password="$OF_PASSWORD"
+kubectl create -f ${tmp_dir}/yaml
+
+cecho "---> Waiting up to ${OF_TIMEOUT} for OpenFaaS gateway to become ready" ${GREEN}
+${TIMEOUT_CMD} ${OF_TIMEOUT} /bin/bash -c 'while [[ $READY -lt 1 ]]; do READY=$(kubectl -n openfaas get deploy gateway -o json | '"${JQBIN}"' .status.readyReplicas); sleep 1; done'
+
+# start kubectl port-forwarding
+cecho "---> Starting port-forwarding to OpenFaaS gateway" ${GREEN}
+kubectl -n openfaas port-forward ${PORT_FWD} >/dev/null &
+
+# get faas-cli
+cecho "---> Downloading faas-cli (${CLI_VERSION})" ${GREEN}
+
+case "${OSTYPE}" in
+darwin*) FAAS_CLI="faas-cli-darwin" ;;
+msys*) FAAS_CLI="faas-cli.exe" ;;
+*) FAAS_CLI="faas-cli" ;;
+esac
+
+curl --silent --show-error --fail -LO https://github.com/openfaas/faas-cli/releases/download/${CLI_VERSION}/${FAAS_CLI}
+chmod +x ./${FAAS_CLI}
+echo ${OF_PASSWORD} | ./${FAAS_CLI} login --password-stdin
+
+# deploy functions used by integration tests
+cecho "---> Downloading templates and installing functions" ${GREEN}
+./${FAAS_CLI} template pull
+./${FAAS_CLI} template pull --overwrite https://github.com/openfaas-incubator/python-flask-template
+./${FAAS_CLI} deploy -f https://raw.githubusercontent.com/embano1/${FAILFN}/master/stack.yml
+./${FAAS_CLI} deploy -f https://raw.githubusercontent.com/embano1/${OKFN}/master/stack.yml
+
+cecho "---> Waiting up to ${FN_TIMEOUT} for ${FAILFN} and ${OKFN} to become ready" ${GREEN}
```

Comment:
Again we can replace timeout logic with closely following when a function is ready as functions are deployments we can use `kubectl rollout status` given that we know the function's name `deployment name == function name`

Reply:
`kubectl` has a `wait` flag to cover this. When I created the script I thought there will be more areas where I would have to wait/timeout (e.g. downloading other bits) and thus standardized on using `timeout`. But I agree that this logic is harder to maintain (vs. `kubectl ... --wait`) and happy to get this changed in another PR.